Fix false positive and false negative cache hits (take 2)

eed3si9n opened the pull request:
Fixes #42
Fixes #45
This a rework of #44 /cc @bjaglin 

I've been studying and documenting `sbt.util.Tracked` in https://github.com/sbt/website/pull/806, and thought this would be a good exercise for me to tackle.

<s>
For `scalafmtCheck` incremental behavior, I've switched to using `Tracked.diffInputs` since last modified stamp should not change during the operation.
</s>

Previously unformatted files are now tracked in a case class `ScalafmtCheckAnalysis`, and is tracked as a previous value of `scalafmtCheckImpl` task. This delays the exception throwing till later, so we have a chance to store them for later use.

For #45, I've added an explicit SHA-1 hash tracker for config file changes. If the configs changes, then we format everything that's passed in.

**Edit**: Adding independent tracker broke the cooperation between the two tasks, so I guess I need to bring that back.

eed3si9n commented on plugin/src/sbt-test/scalafmt-sbt/sbt/test:
```diff
@@ -74,24 +74,6 @@ $ copy-file p13/src/test/scala/Test3.scala p13/src/main/scala/Test3.scala
-$ exec chmod 000 p14/src/main/scala/Test.scala
-######## no-op expected when nothing has changed
-> p14/scalafmtCheck
-> p14/scalafmt
```
Removed p14 because scalafmt here should not be a no-op just because scalafmtCheck moved the tracker forward.

bjaglin replied:
PR commits were squashed when merging so this might not be obvious, but [this was protecting against regressions](https://github.com/scalameta/sbt-scalafmt/commit/e2a0d59546a7d4b23ba41e03caad9efcfcf8fb94) of #39, and should remain relevant despite the implementation change

eed3si9n replied:
ok. I think I understand now. In case `scalfmtCheck` succeeded, I should stop passing `pending` file change since there's nothing to (re)format?

eed3si9n replied:
@poslegm Looking at your error message, it makes sense what's happening. I'll fix to address that.

poslegm commented on plugin/src/main/scala/org/scalafmt/sbt/ScalafmtPlugin.scala:
```diff
@@ -103,19 +123,25 @@ object ScalafmtPlugin extends AutoPlugin {
-    }(sources.toSet).getOrElse(())
+    trackSourcesAndConfig(cacheStoreFactory, sources, config) {
+      (outDiff, configChanged, prev) =>
+        log.debug(outDiff.toString)
```
Missed debug logging?

eed3si9n replied:
Not sure what you mean. This is a debug logging that will be displayed at debug level.

poslegm replied:
Oh, sorry, I didn't pay attention to the logging level

poslegm commented on plugin/src/main/scala/org/scalafmt/sbt/ScalafmtPlugin.scala:
```diff
@@ -146,78 +168,90 @@ object ScalafmtPlugin extends AutoPlugin {
+  ): ScalafmtAnalysis = {
+    trackSourcesAndConfig(cacheStoreFactory, sources, config) {
+      (outDiff, configChanged, prev) =>
+        log.debug(outDiff.toString)
```
Another one

poslegm commented on plugin/src/main/scala/org/scalafmt/sbt/ScalafmtPlugin.scala:
```diff
@@ -233,13 +267,21 @@ object ScalafmtPlugin extends AutoPlugin {
+      scalafmtImpl.value
+      ()
+    },
+    scalafmtImpl := {
```
Is there any way to exclude `*Impl` tasks from public api?

eed3si9n replied:
Because it's not part of the autoImport, it won't be available to build.sbt.
@eatkins suggested using `_` prefix in his plugin to hide the key, I think.

dwijnand replied:
Could you not define these with `Initialize` values defined with `Def.task` and replace:

```scala
(scalafmtImpl / streams).value.cacheStoreFactory
```

with the downstream

```scala
(scalafmt / streams).value.cacheStoreFactory
```
?

eed3si9n replied:
@dwijnand Yea. Now that I'm not using the previous value, maybe I can do that.

bjaglin commented on plugin/src/main/scala/org/scalafmt/sbt/ScalafmtPlugin.scala:
```diff
@@ -57,8 +58,28 @@ object ScalafmtPlugin extends AutoPlugin {
+
   import autoImport._
 
+  case class ScalafmtAnalysis(failed: List[File], pending: List[File])
```
is `pending` ever non-empty ?

eed3si9n replied:
`checkSources` populates it as:

```scala
pending = (prev.pending ++ filesToCheck).distinct
```

so `scalafmt` can pick it up.

eed3si9n replied:
I think I can get rid of pending.

eed3si9n commented:
> Are you planning to look at the task cooperation regression in the near future? If not, I can give it a try!

I'll look at it now, but you're welcome to take my commits and get it to finish too.

eed3si9n commented:
@bjaglin 

> Is it missing the reference to close #42?

It was in the commit message, but I've also moved it to the PR description as well.

> Any reason not to carry over the tests from

Moved the tests. Had to make a minor adjustment to the last one though.

bjaglin commented on plugin/src/main/scala/org/scalafmt/sbt/ScalafmtPlugin.scala:
```diff
@@ -146,78 +163,89 @@ object ScalafmtPlugin extends AutoPlugin {
+          if (configChanged) sources
+          else updatedOrAdded.toList
+        val failed = prev.failedScalafmtCheck filter { _.exists }
+        val files: List[File] = (filesToCheck.toList ++ failed).distinct
```
Extending incrementality to rejected/failed sources (and not just formatted ones) is probably beyond the scope of https://github.com/scalameta/sbt-scalafmt/issues/42, but I guess we don't really need to recheck the files that failed the previous run - running a fresh check against `filesToCheck` only and concatenating the result with the previous analysis (intersected with `outDiff.unmodified`) should do the trick, no? This would require to push up the `log.warn(s"${file.toString} isn't formatted properly!")` though if we want to keep logging the list of files that need to be formatted, no matter the status of the cache.

bjaglin commented on plugin/src/sbt-test/scalafmt-sbt/sbt/test:
```diff
@@ -94,16 +94,38 @@ $ touch p14/src/main/scala/Test.scala
+######## incremental checking expected when all previous validations failed on some sources only
+> p17/test:scalafmt
+$ copy-file changes/bad.scala p17/src/main/scala/Test2.scala
+-> p17/scalafmtCheck
```
It seems that this test is no longer relevant: there is no source in `test` anymore, so the 2 lines above are no-op, and this does not verify the incrementality of `scalafmtCheck`. That process got lost in https://github.com/scalameta/sbt-scalafmt/pull/46/commits/1944e56c87856cee655032e5a71206ecba85d7e4 from what I can see.

The original scripted was formatting the content of `bad.scala` "outside" and copying it back to the same place (to simulate a user update, rather than exercising the project/config tasks themselves which would have biased the test), expecting the check here to run successfully on that updated file without re-running on the `Test.scala` bombed above. I thought using `test/scala/Test2.scala` & `test:scalafmt` as a sandbox was better than checking-in 2 files (one being unformatted, the other one formatted) to show the intent, but it was obviously not clear! What do you think?

eed3si9n replied:
ah. I totally did not see `test:` in `> p17/test:scalafmt`. I think it's better to check in two files.

bjaglin commented on plugin/src/main/scala/org/scalafmt/sbt/ScalafmtPlugin.scala:
```diff
@@ -57,8 +58,19 @@ object ScalafmtPlugin extends AutoPlugin {
+
   import autoImport._
 
+  case class ScalafmtAnalysis(failedScalafmtCheck: List[File])
```
Is the choice of `List` driven by sjsonnew? It looks like it's always handled as a `Set` internally.

poslegm commented on plugin/src/main/scala/org/scalafmt/sbt/ScalafmtPlugin.scala:
```diff
@@ -146,135 +163,167 @@ object ScalafmtPlugin extends AutoPlugin {
-        false
-      },
+  ): ScalafmtAnalysis = {
+    log.info(s"Checking ${sources.size} Scala sources...")
```
It's a little thing, but there's been a changed files count check before:

```
if (changed.size > 0) {
  log.info(s"Checking ${changed.size} Scala sources...")
}
```

So in your version I have these messages:

```
sbt:monixKafka> kafka10/scalafmtCheckAll
[info] Checking 0 Scala sources...
[info] Checking 0 Scala sources...
[success] Total time: 0 s, completed Aug 26, 2019 6:25:58 PM
```

Can we bring `if` block back?

```
if (sources.nonEmpty) {
  log.info(s"Checking ${sources.size} Scala sources...")
}
```

bjaglin commented on plugin/src/sbt-test/scalafmt-sbt/sbt/test:
```diff
@@ -132,6 +132,8 @@ $ copy-file changes/bad.scala p17/src/main/scala/Test2.scala
 -> p17/scalafmtCheck
 # prevent reading the source (without changing the mtime) to detect actual/uncached invocation of scalafmt
 $ exec chmod 000 p17/src/main/scala/Test.scala
+######## incremetnal checking should carry over failure without actually processing the file
```
if we *really* want to assert that the file was not processed, we would have to check that no exception was raised (AFAIK the statement below doesn't differentiate the permission denied caused by the `chmod` from a `false` return value)

bjaglin replied:
sorry I was confused, `scalafmtCheck` raises an exception no matter what (false is never returned), so we would have to assert on the type of exception

poslegm commented:
Shipped 🚀
